File: acwing/算法提高课/第一章动态规划/734.能量石/734.能量石.py
N = 10000 + 5

f = [float("-inf")] * N
f[0] = 0


class stone:
    def __init__(self, s, e, l):
        self.s = s
        self.e = e
        self.l = l

    def __lt__(self, stone):
        return self.s * stone.l < stone.s * self.l


t = int(input())

for k in range(1, t + 1):
    f = [float("-inf")] * N
    f[0] = 0
    n = int(input())
    a = []
    m = 0
    for _ in range(n):
        s, e, l = map(int, input().split())
        a.append(stone(s, e, l))
        m += s
    a.sort()

    for i in range(n):
        s, e, l = a[i].s, a[i].e, a[i].l
        for j in range(m, s - 1, -1):
            f[j] = max(f[j], f[j - s] + e - (j - s) * l)

    res = max(f[:m + 1])

    print(f"Case #{k}: {res}")

File: acwing/算法提高课/第一章动态规划/734.能量石/734.能量石.cpp
#include <algorithm>
#include <cstring>
#include <iostream>
#include <cstdio>

using namespace std;

const int N = 105, M = 10000 + 5;

int f[M];

class stone
{
public:
    int s, e, l;

public:
    bool operator<(const stone &other) const
    {
        return this->s * other.l < this->l * other.s;
    }
};

int main()
{
    int t;
    cin >> t;
    for (int k = 1; k <= t; k++)
    {
        memset(f, -0x3f, sizeof(f));
        f[0] = 0;
        stone a[N];
        int n;
        cin >> n;
        int m = 0;
        for (int j = 0; j < n; j++)
        {
            cin >> a[j].s >> a[j].e >> a[j].l;
            m += a[j].s;
        }
        sort(a, a + n);
        for (int i = 0; i < n; i++)
        {
            int s = a[i].s, e = a[i].e, l = a[i].l;
            for (int j = m; j >= s; j--)
            {
                f[j] = max(f[j], f[j - s] + e - (j - s) * l);
            }
        }
        int res = 0;
        for (int j = 0; j <= m; j++)
            res = max(res, f[j]);
        printf("Case #%d: %d\n", k, res);
    }
    return 0;
}
